### PR TITLE
Rename `delete` to `release` - ready for 0.6.0

### LINKED ISSUE #131: naming: delete vs release
We have `delete`, `deleteIn` and in https://github.com/dart-lang/jnigen/pull/130 `deleteOriginal`.

"delete" refers to deleting the **handle** that holds on to the Java object (`DeleteGlobalRef`).

However, the method is on an object representing a Java object. And we're not actually deleting any Java objects. Instead, we're letting go of the Java object.

Someone using our API does not know about handles, they just know that we are "referring" to a Java object (`JReference`).

So, we should consider renaming "delete" to something that signals we're no longer holding on, no longer referring to the Java object. For example "release". You release the object, the opposite of holding on. (Or maybe the opposite of holding on would be "let go" but that is two words.) Or another option is we "forget" about the Java object, but that seems like we're in control of what we're doing. So, "release"?

Some prior discussion: https://github.com/dart-lang/jnigen/issues/49#issue-1362986168

@mahesh-hegde I want to move more and more away from the JNI implementation details. 🙃
@HosseinYousefi what are your thoughts about this?

### PR DESCRIPTION
**Pull request by HosseinYousefi**
Closes #131 – the final issue of [0.6.0 milestone](https://github.com/dart-lang/jnigen/milestone/4) to get ready for the this month's release.

Also changed the changelog/pubspec to reflect the new 0.6.0 version.

---

Final questions before merging:
* Any alternative to `releaseOriginal`, since it can be verbose?
* Is `releasedIn(arena)` a good name?

### CODE REVIEW

**Comment by dcharkes** [comment]
> * Is `releasedIn(arena)` a good name?

`managedIn` or `releasedBy` maybe?

> * Any alternative to `releaseOriginal`, since it can be verbose?

No good ideas here.

While looking at the code though, I see unnecessary verboseness in some places:

```
      array[0] = "حس".toJString()..deletedIn(arena);
      array[1] = "\$".toJString()..deletedIn(arena);
      array[2] = "33".toJString()..deletedIn(arena);
      expect(array[0].toDartString(deleteOriginal: true), "حس");
      expect(array[1].toDartString(deleteOriginal: true), "\$");
      expect(array[2].toDartString(deleteOriginal: true), "33");
```

This deletes the Java strings twice.

Probably, if there is already an arena, `releaseOriginal` can most likely be ignored and the original object can be using `releasedIn`. It doesn't matter if one executes 20 Dart statements extra before releasing.

Same here:

```
    using((arena) {
      expect(JLong(val).longValue(deleteOriginal: true), val);
      expect((-val).toJLong().longValue(deleteOriginal: true), -val);
    });
```

The `using` can be removed or the JObjects can be `..releasedIn(arena)`.

It's likely a smell if there's an `Arena` and `releaseOriginal` are in the same place.

**Comment by HosseinYousefi** [comment]
> 
> While looking at the code though, I see unnecessary verboseness in some places:

It is unnecessary to delete period, since we have finalizers. However it's not redundant. 

> 
> ```
>       array[0] = "حس".toJString()..deletedIn(arena);
>       array[1] = "\$".toJString()..deletedIn(arena);
>       array[2] = "33".toJString()..deletedIn(arena);
>       expect(array[0].toDartString(deleteOriginal: true), "حس");
>       expect(array[1].toDartString(deleteOriginal: true), "\$");
>       expect(array[2].toDartString(deleteOriginal: true), "33");
> ```
> 
> This deletes the Java strings twice.


No, it doesn't. If it would then we'd get a `DoubleReleaseException`. The array here is a Java array, so `operator []` creates a new global reference every time it's called. Now we could say `(array[0]..deletedIn(arena)).toDartString()` but that seems more verbose.

**Comment by dcharkes** [comment]
> > While looking at the code though, I see unnecessary verboseness in some places:
> 
> It is unnecessary to delete period, since we have finalizers. However it's not redundant.
> 
> > ```
> >       array[0] = "حس".toJString()..deletedIn(arena);
> >       array[1] = "\$".toJString()..deletedIn(arena);
> >       array[2] = "33".toJString()..deletedIn(arena);
> >       expect(array[0].toDartString(deleteOriginal: true), "حس");
> >       expect(array[1].toDartString(deleteOriginal: true), "\$");
> >       expect(array[2].toDartString(deleteOriginal: true), "33");
> > ```
> >   
> > This deletes the Java strings twice.
> 
> No, it doesn't. If it would then we'd get a `DoubleReleaseException`. The array here is a Java array, so `operator []` creates a new global reference every time it's called. Now we could say `(array[0]..deletedIn(arena)).toDartString()` but that seems more verbose.

Oh right, the list is a Java list, so the indexed access creates a new reference. 🤓 

(Side note:
I wonder if we should support assignment into handles to limit the amount of handles:

```dart
var jstring = "حس".toJString();
array[0] = jstring;
// jstring ^= "\$".toJString(); // This is what we use for handles in the C++ code but we can't do that in Dart.
jstring.set("\$".toJString()); // always does deleteOriginal. This doesn't actually reduce the amount of handles. In our C++ code we have the handles and the references you put into handles as two different types.
array[1] = jstring;
// ...
jstring.set(array[0]);
expect(jstring.toDartString(), "حس");
// ...
```

Maybe this approach doesn't really work in Dart because we can't overload operators and use destructors. The destructors free the handles at the end of the syntactic scope when stack allocated.
)

File: README.md [hunk 1]
```diff
@@ -191,10 +191,10 @@ If the errors are similar to `symbol not found`, ensure all dependencies of the
 #### How are classes mapped into bindings?
 Each Java class generates a subclass of `JObject` class, which wraps a `jobject` reference in JNI. Nested classes use `_` as separator, `Example.NestedClass` will be mapped to `Example_NestedClass`.
 
-#### Does `JObject` hold a local or global reference? Does it need to be manually deleted?
+#### Does `JObject` hold a local or global reference? Does it need to be manually released?
 Each Java object returned into Dart creates a JNI global reference. Reference deletion is taken care of by `NativeFinalizer` and that's usually sufficient.
 
-It's a good practice to keep the interface between languages sparse. However, if there's a need to create several references (Eg: in a loop), you can use FFI Arena mechanism (`using` function) and `deletedIn` method, or manually delete the object using `delete` method.
+It's a good practice to keep the interface between languages sparse. However, if there's a need to create several references (Eg: in a loop), you can use FFI Arena mechanism (`using` function) and `releasedIn` method, or manually release the object using `release` method.
```
[the remark this answers is not in the document]
Comment (HosseinYousefi) [reply README.md:197]: `releasedBy` it is.

**Pull request merged**